Lib/gftools/meta.py

```diff
@@ -0,0 +1,76 @@
+"""Generate meta table for a font
+
+This module exports the function "gen_meta_table" which can be used to
+add a meta table designating supported and/or design scripts/languages.
+"""
+from fontTools.ttLib import TTFont, newTable
+import os
+import logging
+import yaml
+import re
+from langcodes import Language
+from langcodes.tag_parser import LanguageTagError
+from langcodes.data_dicts import VALIDITY
+
+__all__ = ["gen_meta_table"]
+
+
+log = logging.getLogger(__name__)
+
+
+def gen_meta_table(ttFont, config):
+    """Generates a meta table from a configuration dictionary.
+
+    Args:
+        ttFont: a TTFont instance
+        config: a dictionary containing ``slng``/``dlng`` keys
+    """
+    assert isinstance(config, dict)
+    if "meta" in ttFont:
+        meta = ttFont["meta"]
+        log.warning("meta table already found; merging configuration values!")
+    else:
+        ttFont["meta"] = meta = newTable("meta")
+    for key in ["slng", "dlng"]:
+        if key not in config:
+            continue
+        values = config[key]
+        if isinstance(values, str):
+            values = re.split(r",\s*", values)
+        if key in meta.data:
+            values.extend(re.split(r",\s*", meta.data[key]))
+        value = _language_list_to_string(values)
+        meta.data[key] = value
+
+
+def _language_list_to_string(values):
+    return ", ".join(
+        sorted([value for value in set(values) if _validate_scriptlangtag(value)])
+    )
+
+
+def _validate_scriptlangtag(value):
+    try:
+        code = Language.get(value)
```

Comment:
`Language.get` by default does some normalization of invalid inputs, as in `eng` to `en`. That means that all of the validation in this function is done in terms of the normalized code, whereas the caller of this function interprets the return value in terms of the original code. Either this function should return the normalized code so the caller can use it, or this line should be `code = Language.get(value, normalize=False)`.